components/raftstore-v2/src/operation/command/write/ingest.rs

```diff
@@ -0,0 +1,123 @@
+    ) -> Result<()> {
+        let ssts = box_try!(ctx.sst_importer.list_ssts());
+        if ssts.is_empty() {
+            return Ok(());
+        }
+        let mut region_ssts: HashMap<_, Vec<_>> = HashMap::default();
+        for sst in ssts {
+            region_ssts
+                .entry(sst.get_region_id())
+                .or_default()
+                .push(sst);
+        }
+        for (region_id, ssts) in region_ssts {
+            if let Err(TrySendError::Disconnected(msg)) = ctx.router.send(region_id, PeerMsg::CleanupImportSst(ssts.into()))
+                && !ctx.router.is_shutdown() {
+                let PeerMsg::CleanupImportSst(ssts) = msg else { unreachable!() };
+                let _ = ctx.schedulers.tablet_gc.schedule(tablet_gc::Task::CleanupImportSst(ssts));
+            }
+        }
+
+        Ok(())
+    }
+}
+
+impl<EK: KvEngine, ER: RaftEngine> Peer<EK, ER> {
+    pub fn on_cleanup_import_sst<T>(
+        &mut self,
+        ctx: &mut StoreContext<EK, ER, T>,
+        ssts: Box<[SstMeta]>,
+    ) {
+        let epoch = self.region().get_region_epoch();
+        let stale_cnt = ssts
+            .iter()
+            .filter(|sst| util::is_epoch_stale(sst.get_region_epoch(), epoch))
+            .count();
+        if stale_cnt == 0 {
+            return;
+        }
+        let stale_ssts = if stale_cnt == ssts.len() {
+            ssts
+        } else {
+            let mut ssts = Vec::from(ssts);
+            ssts.retain(|sst| util::is_epoch_stale(sst.get_region_epoch(), epoch));
+            ssts.into_boxed_slice()
+        };
+        let _ = ctx
+            .schedulers
+            .tablet_gc
+            .schedule(tablet_gc::Task::CleanupImportSst(stale_ssts));
+    }
+}
+
+impl<EK: KvEngine, R: ApplyResReporter> Apply<EK, R> {
+    #[inline]
+    pub fn apply_ingest(&mut self, ssts: Vec<SstMeta>) -> Result<()> {
+        PEER_WRITE_CMD_COUNTER.ingest_sst.inc();
+        let mut infos = Vec::with_capacity(ssts.len());
+        for sst in &ssts {
+            if let Err(e) = check_sst_for_ingestion(sst, self.region()) {
+                error!(
+                    self.logger,
+                    "ingest fail";
+                    "sst" => ?sst,
+                    "region" => ?self.region(),
+                    "error" => ?e
+                );
+                let _ = self.sst_importer().delete(sst);
+                return Err(e);
+            }
+            match self.sst_importer().validate(sst) {
+                Ok(meta_info) => infos.push(meta_info),
+                Err(e) => {
+                    slog_panic!(self.logger, "corrupted sst"; "sst" => ?sst, "error" => ?e);
+                }
+            }
+        }
+        // Unlike v1, we can't batch ssts accross regions.
+        self.flush();
+        if let Err(e) = self.sst_importer().ingest(&infos, self.tablet()) {
+            slog_panic!(self.logger, "ingest fail"; "ssts" => ?ssts, "error" => ?e);
```

Comment:
Why panic?

Reply:
v1 panics so here inherits the behavior.